JSON results / barrett15 false positive fix / version bump

brubsby opened the pull request:
I've been working on making changes to mfakto that I've made to mfaktc to support json results, as part of an effort to integrate mfakt* programs with primenet.py, as part of this I've also fixed a false negative issue with cards using double precision math for divisions, and am hoping to release prerelease builds to others for testing.

It seems like this repo hasn't kept up putting out releases, but that would be helpful to me. I can release builds from my fork, but I think bdot is not active anymore, so I'm not sure to what extent I can contribute to this repo.

Also release 0.17pre7 on this repo seems very bizarre, the actual version number of the packaged software is like 0.15pre7 (or something). Could that release be renamed to 0.15pre7? It's caused some confusion in the mfakto thread on mersenneforum

brubsby committed:
changes from mfaktc 0.23

brubsby committed:
fix several bugs

brubsby committed:
fix incorrect factors in json

brubsby committed:
attempt memset fix

brubsby committed:
fix memset error

brubsby committed:
fix barrett15.cl division false negative bug, fix logging, fix typos

brubsby committed:
bump readme version

brubsby committed:
small feedback fixes

proski commented on src/checkpoint.h:
```diff
@@ -17,6 +17,6 @@ You should have received a copy of the GNU General Public License
-void checkpoint_write(unsigned int exp, int bit_min, int bit_max, unsigned int cur_class, int num_factors);
-int checkpoint_read(unsigned int exp, int bit_min, int bit_max, unsigned int *cur_class, int *num_factors, int verbosity);
+void checkpoint_write(unsigned int exp, int bit_min, int bit_max, unsigned int cur_class, int num_factors, char* factors_string, unsigned long long int bit_level_time);
+int checkpoint_read(unsigned int exp, int bit_min, int bit_max, unsigned int *cur_class, int *num_factors, char* factors_string, unsigned long long int bit_level_time, int verbosity);
```
There is a discrepancy between the header and the implementation in the type of `bit_level_time`. The header wants a scalar, the implementation wants a pointer. Please use a strict modern compiler and fix the warnings.

brubsby replied:
This was likely a typo on my part, my apologies. MSC didn't seem to give a warning on build in visual studio, not sure if this is due to the project settings or what. But it did have a barely noticeable green ribbon on the actual code here. Will be fixed in the next commit.

proski commented on src/read_config.c:
```diff
@@ -18,7 +18,13 @@ along with mfaktc (mfakto).  If not, see <http://www.gnu.org/licenses/>.
 
 #include <stdio.h>
+#ifdef _WIN32
 #include <string.h>
```
`string.h` should be included unconditionally, it's needed for `strncpy`.

brubsby replied:
My understanding was that strings.h was the equivalent of string.h for windows, but only mainly differed in the naming of the case insensitive string comparison. This seemed better than rolling our own portable string comparison, but perhaps my info was outdated.

proski commented on src/read_config.c:
```diff
@@ -18,7 +18,13 @@ along with mfaktc (mfakto).  If not, see <http://www.gnu.org/licenses/>.
 #include <string.h>
+#define strcasecmp _stricmp
+#else // assuming POSIX or BSD compliant system
+#include <strings.h>
```
`strings.h` should be avoided until absolutely necessary. I can comment this out on Linux and this file still compiles.

brubsby replied:
My understanding for this was that it was necessary for case insensitive string comparisons on non windows systems. Perhaps I was operating under bad assumptions though.

proski replied:
`strings.h` is not a replacement for `string.h`, at least not on Linux with glibc. For instance, including `strings.h` without `string.h` makes `strlen`, `strcpy` and other functions undeclared and break compilation.

`strcasecmp` was indeed introduced in `strings.h`, but glibc makes it available in `string.h`. I can confirm that using `string.h` instead of `strings.h` would not break compilation on MacOS if `strcasecmp` is used (`string.h` includes `strings.h`). I would not worry too much about other non-glibc systems.

proski commented on src/mfaktc.c:
```diff
@@ -612,24 +624,25 @@ other return value
   if(mystuff->mode == MODE_NORMAL)
   {
-    if((mystuff->checkpoints > 0) && (checkpoint_read(mystuff->exponent, mystuff->bit_min, mystuff->bit_max_stage, &cur_class, &factorsfound, mystuff->verbosity) == 1))
+    if((mystuff->checkpoints > 0) && (checkpoint_read(mystuff->exponent, mystuff->bit_min, mystuff->bit_max_stage, &cur_class, &factorsfound, mystuff->factors_string, &(mystuff->stats.bit_level_time), mystuff->verbosity) == 1))
```
This doesn't compile on Linux, as `mystuff->stats.bit_level_time` is `unsigned long`, but the expected argument is `unsigned long long`. One way to fix it is to add a temporary variable of the expected type and then copy the value to `mystuff->stats.bit_level_time`.

brubsby replied:
I had perhaps a bit naively ported over changes to mfaktc to here, and this is one point where they diverged, but MSC 19.40.33811 didn't complain, so I assumed it was perhaps okay as there was no error or warning.

mystuff in mfakto uses cl_ulong, which I don't fully understand the purpose of, but I figured I should continue the pattern. a ulong number of milliseconds is 5e8 years, so we shouldn't need a ulong long for this variable. I'll attempt to just change the type to ulong throughout.

brubsby replied:
Delving deeper it seems like cl_ulong is defined as:
`typedef unsigned __int64        cl_ulong;`
which is the minimum size allocated to a c `unsigned long long` iirc.

So I guess we don't actually need to change either of the types, we only need to translate between them more smoothly like you say.

I am a bit surprised your compiler is saying that it's an unsigned long though. As we define it as a cl_ulong in my_types.h:
`cl_ulong bit_level_time;            /* time (in ms) since the bitlevel was started */`

brubsby replied:
Does this mean your compiler is only using a 4 byte unsigned int for `cl_ulong`? but mine is using an 8 byte unsigned int for it?

proski replied:
Here's the exact error message on 64-bit Linux:

```
mfaktc.c: In function ‘tf’:
mfaktc.c:639:168: error: passing argument 7 of ‘checkpoint_read’ from incompatible pointer type [-Wincompatible-pointer-types]
  639 |     if((mystuff->checkpoints > 0) && (checkpoint_read(mystuff->exponent, mystuff->bit_min, mystuff->bit_max_stage, &cur_class, &factorsfound, mystuff->factors_string, &(mystuff->stats.bit_level_time), mystuff->verbosity) == 1))
      |                                                                                                                                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                                                                                                                        |
      |                                                                                                                                                                        cl_ulong * {aka long unsigned int *}
In file included from mfaktc.c:40:
checkpoint.h:21:154: note: expected ‘long long unsigned int *’ but argument is of type ‘cl_ulong *’ {aka ‘long unsigned int *’}
   21 | int checkpoint_read(unsigned int exp, int bit_min, int bit_max, unsigned int* cur_class, int *num_factors, char* factors_string, unsigned long long int* bit_level_time, int verbosity);
      |                                                                                                                                  ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
```

Just because pointers point to variables of the same size it doesn't make them compatible. It should be the same type.

The gcc documentation says that `-Wincompatible-pointer-types` is an error for C99 and newer. I'm using gcc 14.2.1 that defaults to C17.

proski commented:
The message "This branch has no conflicts with the base branch" is misleading. There are conflicts, but GitHub fails to see them. One of the commits is a merge commit. It should normally be avoided in pull requests. Non-linear history is harder to bisect in case of problems.

proski commented on src/mfakto.cpp:
```diff
@@ -136,6 +136,7 @@ int init_CLstreams(int gs_reinit_only)
         printf("ERROR: malloc(h_ktab[%d]) failed\n", i);
         return 1;
       }
+      memset(mystuff.h_ktab[i], 0, sizeof(mystuff.h_ktab[i]));
```
This is wrong. If you want to zeroize the pointer, you can do
```
memset(&mystuff.h_ktab[i], 0, sizeof(mystuff.h_ktab[i]));
```
If you hand to zeroize the memory it points to, use
```
memset(mystuff.h_ktab[i], 0, sizeof(*mystuff.h_ktab[i]));
```

brubsby replied:
Good catch! I want to zeroize the memory it's pointing to.

proski commented on src/mfakto.cpp:
```diff
@@ -152,6 +153,7 @@ int init_CLstreams(int gs_reinit_only)
       printf("ERROR: malloc(h_RES) failed\n");
       return 1;
     }
+    memset(mystuff.h_RES, 0, sizeof(mystuff.h_RES));
```
This is using a pointer size to zeroize the data it points to. Modern gcc (14.2.1 in my case) flags it as a warning. This should probably be `memset(mystuff.h_RES, 0, sizeof(*mystuff.h_RES));`

brubsby replied:
Thanks for this catch, it wasn't flagged as a warning with MSC 19.40.33811.

ixfd64 replied:
I also got some warnings on macOS, but your latest changes resolved them.

brubsby commented:
> The message "This branch has no conflicts with the base branch" is misleading. There are conflicts, but GitHub fails to see them. One of the commits is a merge commit. It should normally be avoided in pull requests. Non-linear history is harder to bisect in case of problems.

I had attempted to do a rebase, but I think accidentally did a "git pull" before committing my resolution to the merge conflicts for the rebase. If you have any specific instructions for how I might undo the merge and have it be solely a rebase, I will certainly perform them and force push them to my branch. I've also allowed edits to the PR now if it's possible and desired for a contributor to do them on my behalf.

brubsby committed:
address various PR feedback

brubsby commented:
> The message "This branch has no conflicts with the base branch" is misleading. There are conflicts, but GitHub fails to see them. One of the commits is a merge commit. It should normally be avoided in pull requests. Non-linear history is harder to bisect in case of problems.

I've attempted to rebase the merge out of existence, hopefully this hasn't caused any other problems, let me know if there's anything else I can do.

brubsby committed:
add RX 550 to auto recognition

proski committed:
Fix compile errors on Linux

Change type of `bit_level_time` to match the functions that process it.

Include `string.h` unconditionally. Don't include `strings.h`.

proski commented:
It there any test case that would fail without the barrett15 fix (with double precision) and pass with the fix? It should be added to catch similar issues in the future.

brubsby commented:
> It there any test case that would fail without the barrett15 fix (with double precision) and pass with the fix? It should be added to catch similar issues in the future.

IIRC, there were many test cases in -st and -st2, though maybe some of them could be added to the quickstart tests.

k3ack3r commented:
Running on windows, 6700XT - fails the sanity selfcheck for me.

```
Compiling kernels.
  GPUSievePrimes (adjusted) 81206
  GPUsieve minimum exponent 1037054
Started a simple self-test ...
ERROR: self-test failed for M51332417 (cl_barrett15_69_gs)
  no factor found
ERROR: self-test failed for M50896831 (cl_barrett15_71_gs)
  no factor found
ERROR: self-test failed for M50979079 (cl_barrett15_73_gs)
  no factor found
ERROR: self-test failed for M51232133 (cl_barrett15_73_gs)
  no factor found
ERROR: self-test failed for M50830523 (cl_barrett15_73_gs)
  no factor found
ERROR: self-test failed for M50752613 (cl_barrett15_73_gs)
  no factor found
ERROR: self-test failed for M51507913 (cl_barrett15_73_gs)
  no factor found
ERROR: self-test failed for M51916901 (cl_barrett15_74_gs)
  no factor found
ERROR: self-test failed for M50805581 (cl_barrett15_82_gs)
  no factor found
ERROR: self-test failed for M51157429 (cl_barrett15_82_gs)
  no factor found
ERROR: self-test failed for M51406151 (cl_barrett15_82_gs)
  no factor found
ERROR: self-test failed for M51478381 (cl_barrett15_82_gs)
  no factor found
ERROR: self-test failed for M51350527 (cl_barrett15_82_gs)
  no factor found
ERROR: self-test failed for M53061139 (cl_barrett15_82_gs)
  no factor found
ERROR: self-test failed for M48629519 (cl_barrett15_83_gs)
  no factor found
ERROR: self-test failed for M55069117 (cl_barrett15_69_gs)
  no factor found
ERROR: self-test failed for M45448679 (cl_barrett15_83_gs)
  no factor found

Self-test statistics
  number of tests           30
  successful tests          13
  no factor found           17

self-test FAILED!


```

ixfd64 commented:
@k3ack3r Does this happen with the latest "stable" code?
